timemachine/cpp/src/bound_potential.cu

```diff
@@ -36,22 +36,35 @@ void BoundPotential::execute_host(
     d_x.copy_from(h_x);
     d_box.copy_from(h_box);
 
-    DeviceBuffer<unsigned long long> d_du_dx(N * D);
-    DeviceBuffer<__int128> d_u(1);
+    std::unique_ptr<DeviceBuffer<unsigned long long>> d_du_dx(nullptr);
```

Comment:
Can you explain the rationale for the `unique_ptr` here? Not sure why we need the extra level of indirection here? Performance reasons?

Reply:
The rationale here is that we don't allocate GPU memory if we don't need it. The reduction in memory allocations is where most of the performance improvement comes from.